numpy/ctypeslib.py

```diff
@@ -319,120 +319,44 @@ def ndpointer(dtype=None, ndim=None, shape=None, flags=None):
 
+def _get_typecodes():
+    """ Get a dictionary mapping __array_interface__ formats to ctypes types """
+    ct = ctypes
     simple_types = [
-        ((ct.c_byte, ct.c_short, ct.c_int, ct.c_long, ct.c_longlong), "i"),
-        ((ct.c_ubyte, ct.c_ushort, ct.c_uint, ct.c_ulong, ct.c_ulonglong), "u"),
-        ((ct.c_float, ct.c_double), "f"),
+        ct.c_byte, ct.c_short, ct.c_int, ct.c_long, ct.c_longlong,
+        ct.c_ubyte, ct.c_ushort, ct.c_uint, ct.c_ulong, ct.c_ulonglong,
+        ct.c_float, ct.c_double,
     ]
 
-    # Prep that numerical ctypes types:
-    for types, code in simple_types:
-        for tp in types:
-            prep_simple(tp, "%c%d" % (code, ct.sizeof(tp)))
-
-    ################################################################
-    # array types
-
-    _ARRAY_TYPE = type(ct.c_int * 1)
-
-    def prep_array(array_type):
-        """Given a ctypes array type, construct and attach an
-        __array_interface__ property to it if it does not yet have one.
-        """
-        try: array_type.__array_interface__
-        except AttributeError: pass
-        else: return
-
-        shape = []
-        ob = array_type
-        while type(ob) is _ARRAY_TYPE:
-            shape.append(ob._length_)
-            ob = ob._type_
-        shape = tuple(shape)
-        ai = ob().__array_interface__
-        descr = ai['descr']
-        typestr = ai['typestr']
-
-        def __array_interface__(self):
-            return {'descr': descr,
-                    '__ref': self,
-                    'strides': None,
-                    'shape': shape,
-                    'version': 3,
-                    'typestr': typestr,
-                    'data': (ct.addressof(self), False),
-                    }
-
-        array_type.__array_interface__ = property(__array_interface__)
-
-    def prep_pointer(pointer_obj, shape):
-        """Given a ctypes pointer object, construct and
-        attach an __array_interface__ property to it if it does not
-        yet have one.
-        """
-        try: pointer_obj.__array_interface__
-        except AttributeError: pass
-        else: return
+    return {
+        _dtype(ctype).str: ctype
+        for ctype in simple_types
+    }
 
-        contents = pointer_obj.contents
-        dtype = _dtype(type(contents))
 
-        inter = {'version': 3,
-                 'typestr': dtype.str,
-                 'data': (ct.addressof(contents), False),
-                 'shape': shape}
+def _ctype_ndarray(el_type, shape):
+    """ Create an ndarray of the given element type and shape """
+    for dim in shape[::-1]:
+        el_type = el_type * dim
+    return el_type
 
-        pointer_obj.__array_interface__ = inter
 
-    ################################################################
-    # public functions
+if ctypes is not None:
+    _typecodes = _get_typecodes()
 
-    def as_array(obj, shape=None):
+    def as_array(obj, shape=()):
```

Comment:
it would be safer to have `shape=None` and in the body of the code `if shape is None: shape = ()` 
While in the current version `shape` is read-only, this might future-proof the function

Reply:
Safer from what perspective?

Reply:
whoops, shape is a zero-length-tuple so immutable. Never mind, sorry for the noise.

Reply:
You've made me realize that defaulting to None and erroring is probably closer to the old behaviour

Reply:
Updated to reject `shape=None`